docs/posts/wow-analysis/index.qmd

```diff
@@ -0,0 +1,174 @@
+        ibis.window(group_by=wowah_data.columns, order_by=wowah_data.timestamp)
+    )
+).filter(_.row_num > 0)
+
+wowah_duplicates.count()
+```
+
+I suspect this data was captured by a single player spamming “/who” in the game, most likely using an AddOn, about every ten minutes. It’s possible the same players could have been captured twice, depending on how the command was being filtered.
+
+I’ll remove these duplicates.
+
+```{python}
+wowah_data = wowah_data.distinct()
+```
+
+### Which player logged in the most?
+We just mentioned that there was a single player likely capturing these results. Let’s find out who that is.
+
+```{python}
+(
+    wowah_data
+    .group_by([_.char, _.race, _.charclass])
+    .agg(sessions=_.count())
+    .order_by(_.sessions.desc())
+)
+```
+
+That Troll Hunter that never exceeded level 1 is likely our person with 42,770 sessions.
+
+### Who leveled the fastest from 70–80?
+At the end of the year, there were 884 level 80s. Which one leveled the fastest?
+
+This feels like one of those “trying to trick” you types of questions, and there are many solutions to it. This will involve filtering, grouping, and aggregating to compute each character's time taken to level from 70 to 80.
+
+Let’s start by creating an expression to filter to only the level 80 characters, then join it to filter and identify only where they were level 70 or 80. We’re only concerned with three columns so that we will select only those.
+
+```{python}
+max_level_chars = wowah_data.filter(_.level == 80).select(_.char).distinct()
+wowah_data_filtered = (
+    wowah_data
+    .join(max_level_chars, "char", how="inner")
+    .filter(_.level.isin([70, 80]))
+    .select(_.char, _.level, _.timestamp)
+)
+
+wowah_data_filtered
+```
+
+```{python}
+level_calc = (
+    wowah_data_filtered.group_by(["char"])
+    .mutate(
+        ts_70=_.timestamp.max(where=_.level == 70),
+        ts_80=_.timestamp.min(where=_.level == 80),
+    )
+    .drop(["level", "timestamp"])
+    .distinct()
+    .mutate(days_from_70_to_80=(_.ts_80.delta(_.ts_70, "day")))
+    .order_by(_.days_from_70_to_80)
+)
+
+level_calc
+```
+The data is filtered and grouped by character, and two new columns are created to represent timestamps for levels 70 and 80. The data is then cleaned, and a new column is created to calculate the time taken to level from 70 to 80. Finally, the dataset is sorted by this column.
+
+This isn’t perfect, as I found a case where there was a player who seemed to have quit in March and then returned for the new expansion. They hit 71 before it looks like their login at 70 was captured later. If you’re curious, take a look at **char=21951** for yourself.
+
+### How did they level?
+
+Let’s grab all the details from the previous result and join it back to get the timestamp and zone data.
+
+```{python}
+leveler_zones = (
+    level_calc.join(wowah_data, "char", how="inner")
+    .filter((_.timestamp >= _.ts_70) & (_.timestamp <= _.ts_80))
```

Comment:
```suggestion
    .filter(_.timestamp.between(_.ts_70, _.ts_80 >= _.ts_70))
```

If you want, up to you.

Reply:
Thank you for this! I prefer the `between`.